File: Algorithms/138. Copy List with Random Pointer/golang/solution.iterative.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Next *Node
 *     Random *Node
 * }
 */

func copyRandomList(head *Node) *Node {
	visited := make(map[*Node]*Node)
	for node := head; node != nil; node = node.Next {
		visited[node] = &Node{Val: node.Val}
	}
	for node := head; node != nil; node = node.Next {
		cur := visited[node]
		cur.Next, cur.Random = visited[node.Next], visited[node.Random]
	}
	return visited[head]
}

File: Algorithms/138. Copy List with Random Pointer/javascript/solution.iterative.js
/**
 * // Definition for a Node.
 * function Node(val, next, random) {
 *    this.val = val;
 *    this.next = next;
 *    this.random = random;
 * };
 */

/**
 * @param {Node} head
 * @return {Node}
 */

const visited = new Map();

const getClonedNode = (node) => {
    if (!node) return null;
    if (visited.has(node)) return visited.get(node);
    const newNode = new Node(node.val);
    visited.set(node, newNode);
    return newNode;
};

var copyRandomList = function (head) {
    if (!head) return head;

    let oldNode = head,
        newNode = new Node(oldNode.val);
    visited.set(oldNode, newNode);

    while (oldNode) {
        newNode.random = getClonedNode(oldNode.random);
        newNode.next = getClonedNode(oldNode.next);
        oldNode = oldNode.next;
        newNode = newNode.next;
    }
    return visited.get(head);
};
